src/Microsoft.AspNet.Mvc.Razor/Precompilation/CodeAnalysisSymbolBasedTypeInfo.cs

```diff
@@ -0,0 +1,251 @@
+        public IEnumerable<IPropertyInfo> Properties
+        {
+            get
+            {
+                if (_properties == null)
+                {
+                    _properties = PopulateProperties();
+                }
+
+                return _properties;
+            }
+        }
+
+        /// <inheritdoc />
+        public IEnumerable<TAttribute> GetCustomAttributes<TAttribute>()
+            where TAttribute : Attribute
+        {
+            return CodeAnalysisAttributeUtilities.GetCustomAttributes<TAttribute>(_type, _symbolLookup);
+        }
+
+        /// <inheritdoc />
+        public bool ImplementsInterface(System.Reflection.TypeInfo interfaceType)
+        {
+            var interfaceSymbol = _symbolLookup.GetSymbol(interfaceType);
+            return _type.AllInterfaces.Any(implementedInterface => implementedInterface == interfaceSymbol);
+        }
+
+        /// <inheritdoc />
+        public string[] GetGenericDictionaryParameterNames()
+        {
+            var dictionarySymbol = _symbolLookup.GetSymbol(OpenGenericDictionaryTypeInfo);
+
+            INamedTypeSymbol dictionaryInterface;
+            if (_type.Kind == SymbolKind.NamedType &&
+                ((INamedTypeSymbol)_type).ConstructedFrom == dictionarySymbol)
+            {
+                dictionaryInterface = (INamedTypeSymbol)_type;
+            }
+            else
+            {
+                dictionaryInterface = _type
+                    .AllInterfaces
+                    .FirstOrDefault(implementedInterface => implementedInterface.ConstructedFrom == dictionarySymbol);
+            }
+
+            if (dictionaryInterface != null)
+            {
+                Debug.Assert(dictionaryInterface.TypeArguments.Length == 2);
+
+                return new[]
+                {
+                    GetFullName(dictionaryInterface.TypeArguments[0]),
+                    GetFullName(dictionaryInterface.TypeArguments[1])
+                };
+            }
+
+            return null;
+        }
+
+        /// <summary>
+        /// Gets the assembly qualified named of the specified <paramref name="symbol"/>.
+        /// </summary>
+        /// <param name="symbol">The <see cref="ITypeSymbol" /> to generate the name for.</param>
+        /// <returns>The assembly qualified name.</returns>
+        public static string GetAssemblyQualifiedName([NotNull] ITypeSymbol symbol)
+        {
+            var builder = new StringBuilder();
+            GetAssemblyQualifiedName(builder, symbol);
+
+            return builder.ToString();
+        }
+
+        private List<IPropertyInfo> PopulateProperties()
+        {
+            var properties = new List<IPropertyInfo>();
+            var overridenProperties = new HashSet<IPropertySymbol>();
+            var type = _type;
+
+            while (type != null)
+            {
+                foreach (var member in type.GetMembers())
+                {
+                    if (member.Kind == SymbolKind.Property)
```

Comment:
Note (doesn't matter one way or the other): Could do a where clause to reduce nesting 1 level.